### PR TITLE
Skip execution of analyzers reporting info and hidden diagnostics dur…

### LINKED ISSUE #42166: Support analyzers that are disabled by default in command line build but enabled in other hosts
### Current state
Today, every analyzer that is installed via NuGet package and marked as enabled by default runs in both command line build and IDE, regardless of the analyzer severity. This is completely reasonable as almost all the analyzers are designed to be warnings by default that can break build. Users have to explicitly install the analyzer NuGet package, so any command line build impact happens as part of user action of installing the analyzer.

### Feature request
We are planning to include .NET5 analyzers and possibly even IDE code style analyzers in the .NET SDK (see [this](https://github.com/dotnet/roslyn/issues/38480#issuecomment-578903218) proposed experience for code style analyzers for details). Only a small selected set of analyzers from the .NET analyzers package can be enabled by default as warnings, and all the code style ones have to be suggestions or hidden. To improve discoverability of analyzers in .NET5 analyzers package and supporting IDE code style suggestions, it is imperative that we support analyzers/descriptors to mark themselves as being disabled by default in command line build, which get enabled by explicit user configuration via .editorconfig. This ensure we can keep delivering value to customers with new analyzers lighting up in IDE as suggestions, without impacting command line build by default to prevent any functional compat or performance issues.

### Proposal
I can think of two potential approaches for implementing this feature in the analyzer driver:

1. Add a new overload to `DiagnosticDescriptor` [constructor](http://sourceroslyn.io/#Microsoft.CodeAnalysis/Diagnostic/DiagnosticDescriptor.cs,81), which takes an additional flag, say `bool isEnabledByDefaultInCommandLine`, apart from the already existing `bool isEnabledByDefault`. I personally don't like this proposal as it unnecessarily muddlles up public API surface for a scenario that is very likely not going to be used by majority of analyzer authors, if any.
2. Add a new custom tag, say `DisabledByDefaultOnCommandLine`, to [WellKnownDiagnosticTags](http://sourceroslyn.io/#Microsoft.CodeAnalysis/Diagnostic/WellKnownDiagnosticTags.cs,03ef56ba68ffca36), and the analyzer driver will ensure that from command line compiler context, it also considers this custom tag in deciding if a particular analyzer/descriptor is disabled.

I personally prefer the second approach as it does not have any significant public API impact, while also allowing us to add support for this scenario. I implemented a prototype for it, and with this functionality enabled, adding a NuGet package reference to the [current candidate .NET5 analyzers package](https://dotnet.myget.org/feed/roslyn-analyzers/package/nuget/Microsoft.CodeAnalysis.NetAnalyzers), which has bunch of enabled suggestions but no enabled warning, leads to bunch of new CA suggestions in the IDE with absolutely no command line build impact in terms of functionality and performance.

### PR DESCRIPTION
**Pull request by mavasani**
…ing build

Closes #42166. Implements the design from https://github.com/dotnet/roslyn/issues/42166#issuecomment-617435399.

This PR is aimed towards improving the build performance in presence of analyzers that report Info (Suggestion) and/or Hidden diagnostics. Hidden diagnostics are never reported in build and Info diagnostics are only officially supported in presence of /errorlog, we will retain this support. The command line compilers do report Info diagnostics on command line today, but these are filtered away by MSBuild and are effectively never visible to users. This makes this a non-breaking scenario in .NET core build environment and an extreme corner case breaking change on desktop (clients relying on info diagnostics from raw csc/vbc output). Note that such clients would still have a workaround to get the prior behavior by enabling /errorlog. This change has been discussed with Jared and Andy, who are fine with the new semantics.

This optimization should improve build performance when users install the new IDE CodeStyle NuGet package (mostly hidden diagnostics) and the upcoming .NET5 analyzers package (has large number of info/hidden diagnostics).

### CODE REVIEW

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.AnalyzerConfigOptionNames.cs [hunk 1]
```diff
@@ -1,21 +0,0 @@
-﻿// Licensed to the .NET Foundation under one or more agreements.
-// The .NET Foundation licenses this file to you under the MIT license.
-// See the LICENSE file in the project root for more information.
-
-namespace Microsoft.CodeAnalysis.Diagnostics
-{
-    internal partial class AnalyzerDriver
-    {
-        private static class AnalyzerConfigOptionNames
```
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.AnalyzerConfigOptionNames.cs:9]: All members moved to AnalyzerConfigExtensions.cs

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs [hunk 2]
```diff
@@ -1642,79 +1655,59 @@ private bool TryProcessCompilationCompleted(CompilationCompletedEvent endEvent,
...
             var tree = diagnostic?.Location.SourceTree;
-            if (tree == null || analyzerOptions == null || diagnostic.CustomTags.Contains(WellKnownDiagnosticTags.Compiler) || diagnostic.IsNotConfigurable())
+            if (tree != null &&
+                analyzerOptions.TryGetSeverityFromBulkConfiguration(tree, compilation, diagnostic.Descriptor, out ReportDiagnostic severity))
             {
-                return diagnostic;
+                diagnostic = diagnostic.WithReportDiagnostic(severity);
             }
 
-            // If user has explicitly configured severity for this diagnostic ID, that should be respected and
```
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs:1692]: All deleted code here is moved as-is to a shared extension method on AnalyzerOptions in AnalyzerOptionsExtensions.cs

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs [hunk 3]
```diff
@@ -925,16 +925,23 @@ private int RunCore(TextWriter consoleOutput, ErrorLogger errorLogger, Cancellat
                         additionalFileAnalyzerOptions);
                 }
 
-                Diagnostics.AnalyzerOptions analyzerOptions = CreateAnalyzerOptions(
+                AnalyzerOptions analyzerOptions = CreateAnalyzerOptions(
                     additionalTextFiles, analyzerConfigProvider);
 
+                // Always filter out 'Hidden' analyzer diagnostics in build.
+                // Filter out 'Info' analyzer diagnostics if they are not required to be logged in errorlog.
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs:932]: Can you change this comment to say a bit about "why"? As written it mostly says "what". 

Think it would be good to mention that we're filtering these out because they don't appear in the output of the build hence running them is just spinning CPU cycles for no gain.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs [hunk 4]
```diff
@@ -49,6 +49,11 @@ internal abstract partial class AnalyzerDriver : IDisposable
         /// </summary>
         private readonly bool _hasDiagnosticSuppressors;
 
+        /// <summary>
+        /// Filtered diagnostic severities in the compilation, i.e. diagnostics with effective severity from this set should not be reported.
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs:53]: It's not that they shouldn't be reported though but that they should not be run at all correct?

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs [hunk 5]
```diff
@@ -189,12 +194,17 @@ internal abstract partial class AnalyzerDriver : IDisposable
         /// </summary>
         /// <param name="analyzers">The set of analyzers to include in the analysis</param>
         /// <param name="analyzerManager">AnalyzerManager to manage analyzers for analyzer host's lifetime.</param>
+        /// <param name="filteredSeverities">Filtered diagnostic severities in the compilation, i.e. diagnostics with effective severity from this set should not be reported.</param>
         /// <param name="isComment">Delegate to identify if the given trivia is a comment.</param>
-        protected AnalyzerDriver(ImmutableArray<DiagnosticAnalyzer> analyzers, AnalyzerManager analyzerManager, Func<SyntaxTrivia, bool> isComment)
+        protected AnalyzerDriver(ImmutableArray<DiagnosticAnalyzer> analyzers, AnalyzerManager analyzerManager, ImmutableHashSet<ReportDiagnostic> filteredSeverities, Func<SyntaxTrivia, bool> isComment)
         {
+            Debug.Assert(!filteredSeverities.Contains(ReportDiagnostic.Suppress));
+            Debug.Assert(!filteredSeverities.Contains(ReportDiagnostic.Default));
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs:202]: Reading the PR i'm wondering if we are being overly general here. The API here takes a data type that implies that it can filter out any severity but internally limits it to effectively only two severities. In the code base we only ever provide two combinations of this value. Using a hash table style structure here seems potentially overkill for a dataset that has a max size of two. 

Should we consider instead just making this a simple flags `enum` or two `bool` that filter out the two diagnostic severities that we care about? Or am I missing a scenario here which dictates we should be using the `ImmutableHashSet<T>`?

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs (hunk 3 above)
Comment (mavasani) [reply src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs:932]: Sure, I'll improve the comment.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs (hunk 4 above)
Comment (mavasani) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs:53]: There are 2 effects here:

1. Performance: Do not execute analyzers that only report hidden/info diagnostics.
2. Functional: End users should never see hidden/info diagnostics from the execution.

These overlap for majority of scenarios, except if an analyzer has more then one SupportedDiagnostics, such that one is one warning diagnostic and one is hidden/info diagnostic. We cannot skip analyzer execution due to the warning diagnostic, and still need the post-filtering logic that drops the hidden/info diagnostics reported by this analyzer.

Basically, we can guarantee the functional aspect, and do our best in terms of performance aspect. For majority of analyzers, your statement holds true, but I did not want to add the performance aspect to the comment due to the fact that it cannot always be true.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs (hunk 5 above)
Comment (mavasani) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs:202]: The primary reason to choose a hash set was for simplicity of implementation in this method - we can just do "filteredSeverity.Contains(severity)" check everywhere without worrying about each specific severity. Contains check should be pretty efficient as well.

With 2 bools approach, we need to map each bool to a corresponding severity and have checks such as `if (filterHiddenSeverity && severity == ReportDiagnostic.Hidden || filterInfoSeverity && severity == ReportDiagnostic.Info)`. Similarly, with enum approach, we likely need a flags enum with bit more complex code. I can certainly switch to one of these approaches if you feel that would be more readable and efficient then the current approach.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs (hunk 4 above)
Comment (mavasani) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs:53]: I added an additional "PERF" part to the doc comment.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs (hunk 5 above)
Comment (jaredpar) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs:203]: Hash table still seems heavy here. Did you consider a simple `struct` or `enum` flag that could do the same thing? For instance something like:

```cs
readonly struct HiddenFilter
{
  private readonly int _flag;
  private HiddenFilter(bool includeHidden, bool includeInfo) {
    if (includeHidden) {
      _flag = 1;
    }
    if (includeInfo) {
      _flag |= 2;
    }

    internal static HiddenFilter FilterHidden = new HiddenFilter(includeHidden: false, includeInfo: true);
   
    internal static HiddenFilter FilterbOTH = new HiddenFilter(includeHidden: false, includeInfo: false);

    internal bool Contains(Severity severity) { ... }   
}

```

The simple API can be provided with a lighter weight data structure. Performance typically shows that we should avoid heavier structures like this when we know the item count is below a fixed threshold.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerOptionsExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        private const string SeveritySuffix = "severity";
+
+        private const string DotnetAnalyzerDiagnosticSeverityKey = DotnetAnalyzerDiagnosticPrefix + "." + SeveritySuffix;
+
+        private static string GetCategoryBasedDotnetAnalyzerDiagnosticSeverityKey(string category)
+            => $"{DotnetAnalyzerDiagnosticPrefix}.{CategoryPrefix}-{category}.{SeveritySuffix}";
+
+        /// <summary>
+        /// Tries to get configured severity for the given <paramref name="descriptor"/>
+        /// for the given <paramref name="tree"/> from bulk configuration analyzer config options, i.e.
```
[the remark this answers is not in the document]
Comment (mavasani) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerOptionsExtensions.cs:26]: https://docs.microsoft.com/en-us/visualstudio/code-quality/use-roslyn-analyzers?view=vs-2019#set-rule-severity-of-multiple-analyzer-rules-at-once-in-an-editorconfig-file

It can configure multiple rules simultaneously.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.cs [hunk 7]
```diff
@@ -348,21 +361,37 @@ internal bool IsSupportedDiagnostic(DiagnosticAnalyzer analyzer, Diagnostic diag
...
-                            configuredValue != ReportDiagnostic.Suppress)
+                        // Check if diagnostic is enabled by SyntaxTree.DiagnosticOptions or Bulk configuration from AnalyzerConfigOptions.
+                        if (tree.DiagnosticOptions.TryGetValue(descriptor.Id, out var configuredValue))
+                        {
+                            if (isEnablingSeverity(configuredValue, severityFilter))
+                            {
+                                return true;
+                            }
+                        }
+                        else if (!severityFilter.IsEmpty &&
```
[the remark this answers is not in the document]
Comment (mavasani) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.cs:382]: This is just an optimization as the last check will always fail if there are no filtered severities and we avoid checking for bulk configuration enabling the rule.

[the remark this answers is not in the document]
Comment (mavasani) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.cs:382]: Actually my last statement is not true, I’ll remove the IsEmpty check here.

**Comment by mavasani** [comment]
FYI: https://github.com/dotnet/roslyn/pull/43546/commits/021468a514e6988e7563c8a71c16227c773a9159 Adds unit tests to explicitly guard the core performance optimization in this PR

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs [hunk 8]
```diff
@@ -1642,79 +1656,59 @@ private bool TryProcessCompilationCompleted(CompilationCompletedEvent endEvent,
...
-            var analyzerConfigOptions = analyzerOptions.AnalyzerConfigOptionsProvider.GetOptions(tree);
-
-            // Otherwise, if user has explicitly configured default severity for the diagnostic category, that should be respected.
-            // For example, 'dotnet_analyzer_diagnostic.category-security.severity = error'
-            var categoryBasedKey = AnalyzerConfigOptionNames.GetCategoryBasedDotnetAnalyzerDiagnosticSeverityKey(diagnostic.Category);
-            if (analyzerConfigOptions.TryGetValue(categoryBasedKey, out var value) &&
-                AnalyzerConfigSet.TryParseSeverity(value, out ReportDiagnostic severity))
+            if (!severityFilter.IsEmpty &&
+                diagnostic != null &&
+                severityFilter.Contains(DiagnosticDescriptor.MapSeverityToReport(diagnostic.Severity)))
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs:1709]: What if `diagnostic.Severity` is `Suppress` here? The other place where we run `severityFilter.Contains`  we do an explicit check for `Suppress` as well.

Comment (mavasani) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs:1709]: `Diagnostic.Severity` is actually `DiagnosticSeverity` enum which does not have a `Suppress` field: https://github.com/dotnet/roslyn/blob/d4bf6990e488e58dd97b835284eaf731731df799/src/Compilers/Core/Portable/Diagnostic/DiagnosticSeverity.cs#L9-L35

Other places we are using `ReportDiagnostic`, which has `Suppress` field: https://github.com/dotnet/roslyn/blob/d4bf6990e488e58dd97b835284eaf731731df799/src/Compilers/Core/Portable/Diagnostic/ReportDiagnostic.cs#L9-L43

**Comment by mavasani** [comment]
Thanks @jaredpar. @dotnet/roslyn-compiler for an additional review.

File: src/Compilers/Core/Portable/Diagnostic/SeverityFilter.cs [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    /// <summary>
+    /// Represents a set of filtered diagnostic severities.
+    /// </summary>
+    internal readonly struct SeverityFilter
```
Comment (cston) [inline src/Compilers/Core/Portable/Diagnostic/SeverityFilter.cs:12]: >SeverityFilter [](start = 29, length = 14)

Why a `struct` rather than an `enum` and a `Contains()` extension method?

Comment (mavasani) [reply src/Compilers/Core/Portable/Diagnostic/SeverityFilter.cs:12]: We can also choose a flags enum, just took Jared's suggestion here and went with a simple readonly struct.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs [hunk 10]
```diff
@@ -1642,79 +1656,59 @@ private bool TryProcessCompilationCompleted(CompilationCompletedEvent endEvent,
...
-            }
-
-            var analyzerConfigOptions = analyzerOptions.AnalyzerConfigOptionsProvider.GetOptions(tree);
-
-            // Otherwise, if user has explicitly configured default severity for the diagnostic category, that should be respected.
-            // For example, 'dotnet_analyzer_diagnostic.category-security.severity = error'
-            var categoryBasedKey = AnalyzerConfigOptionNames.GetCategoryBasedDotnetAnalyzerDiagnosticSeverityKey(diagnostic.Category);
-            if (analyzerConfigOptions.TryGetValue(categoryBasedKey, out var value) &&
-                AnalyzerConfigSet.TryParseSeverity(value, out ReportDiagnostic severity))
+            if (!severityFilter.IsEmpty &&
```
Comment (cston) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs:1707]: >!severityFilter.IsEmpty && [](start = 16, length = 26)

Is this check needed?

Comment (mavasani) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs:1707]: Was just avoiding the `MapSeverityToReport` invocation unless required, but that is a cheap method, so I can remove this check.

**Comment by cston** [comment]
>                 var isSuppressed = !diag.IsEnabledByDefault;

Can be moved into `else`.

---
Refers to: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.cs:320 in 021468a. [](commit_id = 021468a514e6988e7563c8a71c16227c773a9159, deletion_comment = False)

File: src/Compilers/Core/Portable/Diagnostic/SeverityFilter.cs [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+                _flag = HiddenBit;
+            }
+
+            if (includeInfo)
+            {
+                _flag |= InfoBit;
+            }
+        }
+
+        internal static SeverityFilter Hidden = new SeverityFilter(includeHidden: true, includeInfo: false);
```
Comment (cston) [inline src/Compilers/Core/Portable/Diagnostic/SeverityFilter.cs:34]: >static [](start = 17, length = 6)

`readonly` for both fields.

File: src/Compilers/Core/Portable/Diagnostic/SeverityFilter.cs [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+    /// Represents a set of filtered diagnostic severities.
+    /// </summary>
+    internal readonly struct SeverityFilter
+    {
+        private const int HiddenBit = 0x01;
+        private const int InfoBit = 0x10;
+
+        private readonly int _flag;
+
+        private SeverityFilter(bool includeHidden, bool includeInfo)
```
Comment (cston) [inline src/Compilers/Core/Portable/Diagnostic/SeverityFilter.cs:19]: >bool includeHidden, bool includeInfo [](start = 31, length = 36)

Since this is a private constructor, consider passing in the bit value directly instead of two bools.

File: src/Compilers/Core/Portable/Diagnostic/SeverityFilter.cs (hunk 9 above)
Comment (cston) [reply src/Compilers/Core/Portable/Diagnostic/SeverityFilter.cs:12]: I'd vote for an `enum` since the type is so simple.

---
In reply to: [414175280](https://github.com/dotnet/roslyn/pull/43546#discussion_r414175280) [](ancestors = 414175280)

Comment (mavasani) [reply src/Compilers/Core/Portable/Diagnostic/SeverityFilter.cs:12]: Sure, let me switch to an enum

**Comment by mavasani** [comment]
> var isSuppressed = !diag.IsEnabledByDefault;
> Can be moved into else.

@cston I avoided that change to retain the ordering of comments.
